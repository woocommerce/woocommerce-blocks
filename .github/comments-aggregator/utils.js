const identifier = `<!-- comments-aggregator -->`;
const separator = '<!-- separator -->';
const footerText =
	'This comment is created by [comments-aggregator](https://github.com/woocommerce/woocommerce-blocks/tree/trunk/.github/comments-aggregator).';
const footer = `\n> <sub>${ footerText }</sub>\n${ identifier }`;

function getSectionId( section ) {
	const match = section.match( /-- section-id: ([^\s]+) --/ );
	return match ? match[ 1 ] : null;
}

function getSectionOrder( section ) {
	const match = section.match( /-- section-order: ([^\s]+) --/ );
	return match ? match[ 1 ] : null;
}

function parseComment( comment ) {
	if ( ! comment ) {
		return [];
	}
	const sections = comment.split( separator );
	return sections
		.map( ( section ) => {
			const sectionId = getSectionId( section );
			const order = getSectionOrder( section );
			/**
			 * This also remove the footer as it doesn't have a section id. This
			 * is intentional as we want the footer to always be the last
			 * section.
			 */
			if ( ! sectionId ) {
				return null;
			}
			return {
				id: sectionId,
				order: parseInt( order, 10 ),
				content: section.trim(),
			};
		} )
		.filter( Boolean );
}

function updateSection( sections, data ) {
	const { sectionId, content, order } = data;
	const index = sections.findIndex( ( section ) => section.id === sectionId );
	const formattedContent = `<!-- section-id: ${ sectionId } -->\n\n<!-- section-order: ${ order } -->\n\n${ content }`;
	if ( index === -1 ) {
		sections.push( {
			id: sectionId,
			content: formattedContent,
		} );
	} else {
		sections[ index ].content = formattedContent;
	}

	return sections;
}

function appendFooter( sections ) {
	return sections.concat( {
		id: 'footer',
		content: footer,
	} );
}

function sortSections( sections ) {
	return sections.sort( ( a, b ) => a.order - b.order );
}

function combineSections( sections ) {
	return sections
		.map( ( section ) => section.content )
		.join( `\n\n${ separator }\n\n` );
}

exports.updateComment = function ( comment, data ) {
	let sections = parseComment( comment );
	sections = updateSection( sections, data );
	sections = sortSections( sections );
	sections = appendFooter( sections );
	return combineSections( sections );
};

exports.isMergedComment = function ( comment ) {
	return (
		comment.body.includes( identifier ) &&
		comment.user.login === 'github-actions[bot]'
	);
};
