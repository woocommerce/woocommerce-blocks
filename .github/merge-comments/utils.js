const separator = '<hr />';
const footer = '<small>This comment is aggregated by merge-comments.</small>';

function getSectionId( section ) {
	const match = section.match( /section-id: ([^\s]+) --/ );
	return match ? match[ 1 ] : null;
}

function parseComment( comment ) {
	if ( ! comment ) {
		return [];
	}
	const sections = comment.split( separator );
	return sections
		.map( ( section ) => {
			const sectionId = getSectionId( section );
			/**
			 * This also remove the footer as it doesn't have a section id. This
			 * is intentional as we want the footer to always be the last
			 * section.
			 */
			if ( ! sectionId ) {
				return null;
			}
			return {
				id: sectionId,
				content: section.trim(),
			};
		} )
		.filter( Boolean );
}

function updateSection( sections, sectionId, content ) {
	const index = sections.findIndex( ( section ) => section.id === sectionId );
	const formattedContent = `<!-- section-id: ${ sectionId } -->\n\n${ content }`;
	if ( index === -1 ) {
		sections.push( {
			id: sectionId,
			content: formattedContent,
		} );
	} else {
		sections[ index ].content = formattedContent;
	}

	return sections;
}

function appendFooter( sections ) {
	return sections.concat( {
		id: 'footer',
		content: `\n${ footer }`,
	} );
}

function combineSections( sections ) {
	return sections
		.map( ( section ) => section.content )
		.join( `\n\n${ separator }\n\n` );
}

exports.updateComment = function ( comment, sectionId, content ) {
	let sections = parseComment( comment );
	sections = updateSection( sections, sectionId, content );
	sections = appendFooter( sections );
	return combineSections( sections );
};

exports.isMergedComment = function ( comment ) {
	return (
		comment.body.includes( footer ) &&
		comment.user.login === 'github-actions[bot]'
	);
};
