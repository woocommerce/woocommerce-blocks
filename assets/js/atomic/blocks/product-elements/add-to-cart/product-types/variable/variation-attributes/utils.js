/**
 * External dependencies
 */
import { keyBy } from 'lodash';

/**
 * Key an array of attributes by name,
 *
 * @param {Object} attributes Attributes array.
 */
export const getAttributes = ( attributes ) => {
	return attributes
		? keyBy(
				Object.values( attributes ).filter(
					( { has_variations: hasVariations } ) => hasVariations
				),
				'name'
		  )
		: [];
};

/**
 * Format variations from the API into a map of just the attribute names and values.
 *
 * @param {Array} variations Variations array.
 */
export const getVariationAttributes = ( variations ) => {
	if ( ! variations ) {
		return {};
	}

	const attributesMap = {};

	variations.forEach( ( { id, attributes } ) => {
		attributesMap[ id ] = attributes.reduce( ( acc, { name, value } ) => {
			acc[ name ] = value;
			return acc;
		}, [] );
	} );

	return attributesMap;
};

/**
 * Given a list of terms, filter them and return options for the select boxes.
 *
 * @param {Object} attributeTerms List of attribute term objects.
 * @param {?Array} validAttributeTerms Valid values if selections have been made already.
 * @return {Array} Value/Label pairs of select box options.
 */
export const getSelectControlOptions = (
	attributeTerms,
	validAttributeTerms = null
) => {
	return Object.values( attributeTerms )
		.map( ( { name, slug } ) => {
			if (
				validAttributeTerms === null ||
				validAttributeTerms.includes( null ) ||
				validAttributeTerms.includes( slug )
			) {
				return {
					value: slug,
					label: name,
				};
			}
			return null;
		} )
		.filter( Boolean );
};

/**
 * Given a list of variations and a list of attribute values, return variations which match.
 *
 * Allows an attribute to be excluded by name. This is used to filter displayed options for
 * individual attribute selects.
 *
 * @param {Object} props
 * @param {Object} props.selectedAttributes List of selected attributes.
 * @param {Object} props.variationAttributes List of variations and their attributes.
 * @param {Object} props.attributeNames List of all possible attribute names.
 * @return {Array} List of matching variation IDs.
 */
export const getVariationsMatchingSelectedAttributes = ( {
	selectedAttributes,
	variationAttributes,
	attributeNames,
} ) => {
	return Object.keys( variationAttributes ).filter( ( variationId ) =>
		attributeNames.every( ( attributeName ) => {
			const selectedAttribute = selectedAttributes[ attributeName ] || '';
			const variationAttribute =
				variationAttributes[ variationId ][ attributeName ];

			// If there is no selected attribute, consider this a match.
			if ( selectedAttribute === '' ) {
				return true;
			}
			// If the variation attributes for this attribute are set to null, it matches all values.
			if ( variationAttribute === null ) {
				return true;
			}
			// Otherwise, only match if the selected values are the same.
			return variationAttribute === selectedAttribute;
		} )
	);
};
