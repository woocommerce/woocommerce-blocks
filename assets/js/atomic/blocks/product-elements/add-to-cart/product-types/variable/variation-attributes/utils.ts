/**
 * External dependencies
 */
import { decodeEntities } from '@wordpress/html-entities';
import {
	Dictionary,
	isObject,
	ProductResponseAttributeItem,
	ProductResponseTermItem,
	ProductResponseVariationsItem,
} from '@woocommerce/types';
import { keyBy } from '@woocommerce/base-utils';

/**
 * Internal dependencies
 */
import { AttributesMap } from '../types';

/**
 * Key an array of attributes by name,
 */
export const getAttributes = (
	attributes?: ProductResponseAttributeItem[] | null
) => {
	return attributes
		? keyBy(
				Object.values( attributes ).filter(
					( { has_variations: hasVariations } ) => hasVariations
				),
				'name'
		  )
		: {};
};

/**
 * Format variations from the API into a map of just the attribute names and values.
 *
 * Note, each item is keyed by the variation ID with an id: prefix. This is to prevent the object
 * being reordered when iterated.
 */
export const getVariationAttributes = (
	/**
	 * List of Variation objects and attributes keyed by variation ID.
	 */
	variations?: ProductResponseVariationsItem[] | null
) => {
	if ( ! variations ) {
		return {};
	}

	const attributesMap: AttributesMap = {};

	variations.forEach( ( { id, attributes } ) => {
		attributesMap[ `id:${ id }` ] = {
			id,
			attributes: attributes.reduce( ( acc, { name, value } ) => {
				acc[ name ] = value;
				return acc;
			}, {} as Dictionary ),
		};
	} );

	return attributesMap;
};

/**
 * Given a list of variations and a list of attribute values, return variations which match.
 *
 * Allows an attribute to be excluded by name. This is used to filter displayed options for
 * individual attribute selects.
 *
 * @return List of matching variation IDs.
 */
export const getVariationsMatchingSelectedAttributes = (
	/**
	 * List of attribute names and terms.
	 *
	 * As returned from {@link getAttributes()}.
	 */
	attributes: Record< string, ProductResponseAttributeItem >,
	/**
	 * Attributes for each variation keyed by variation ID.
	 *
	 * As returned from {@link getVariationAttributes()}.
	 */
	variationAttributes: AttributesMap,
	/**
	 * Attribute Name Value pairs of current selections by the user.
	 */
	selectedAttributes: Record< string, string | null >
) => {
	const variationIds = Object.values( variationAttributes ).map(
		( { id } ) => id
	);

	// If nothing is selected yet, just return all variations.
	if (
		Object.values( selectedAttributes ).every( ( value ) => value === '' )
	) {
		return variationIds;
	}

	const attributeNames = Object.keys( attributes );

	return variationIds.filter( ( variationId ) =>
		attributeNames.every( ( attributeName ) => {
			const selectedAttribute = selectedAttributes[ attributeName ] || '';
			const variationAttribute =
				variationAttributes[ 'id:' + variationId ].attributes[
					attributeName
				];

			// If there is no selected attribute, consider this a match.
			if ( selectedAttribute === '' ) {
				return true;
			}
			// If the variation attributes for this attribute are set to null, it matches all values.
			if ( variationAttribute === null ) {
				return true;
			}
			// Otherwise, only match if the selected values are the same.
			return variationAttribute === selectedAttribute;
		} )
	);
};

/**
 * Given a list of variations and a list of attribute values, returns the first matched variation ID.
 *
 * @return Variation ID.
 */
export const getVariationMatchingSelectedAttributes = (
	/**
	 * List of attribute names and terms.
	 *
	 * As returned from {@link getAttributes()}.
	 */
	attributes: Record< string, ProductResponseAttributeItem >,
	/**
	 * Attributes for each variation keyed by variation ID.
	 *
	 * As returned from {@link getVariationAttributes()}.
	 */
	variationAttributes: AttributesMap,
	/**
	 * Attribute Name Value pairs of current selections by the user.
	 */
	selectedAttributes: Dictionary
) => {
	const matchingVariationIds = getVariationsMatchingSelectedAttributes(
		attributes,
		variationAttributes,
		selectedAttributes
	);
	return matchingVariationIds[ 0 ] || 0;
};

/**
 * Given a list of terms, filter them and return valid options for the select boxes.
 *
 * @see getActiveSelectControlOptions
 *
 * @return Value/Label pairs of select box options.
 */
const getValidSelectControlOptions = (
	/**
	 * List of attribute term objects.
	 */
	attributeTerms: ProductResponseTermItem[],
	/**
	 * Valid values if selections have been made already.
	 */
	validAttributeTerms: Array< string | null > | null = null
) => {
	return Object.values( attributeTerms )
		.map( ( { name, slug } ) => {
			if (
				validAttributeTerms === null ||
				validAttributeTerms.includes( null ) ||
				validAttributeTerms.includes( slug )
			) {
				return {
					value: slug,
					label: decodeEntities( name ),
				};
			}
			return null;
		} )
		.filter( Boolean );
};

/**
 * Given a list of terms, filter them and return active options for the select boxes. This factors in
 * which options should be hidden due to current selections.
 *
 * @return Select box options.
 */
export const getActiveSelectControlOptions = (
	/**
	 * List of attribute names and terms.
	 *
	 * As returned from {@link getAttributes()}.
	 */
	attributes: Record< string, ProductResponseAttributeItem >,
	/**
	 * Attributes for each variation keyed by variation ID.
	 *
	 * As returned from {@link getVariationAttributes()}.
	 */
	variationAttributes: AttributesMap,
	/**
	 * Attribute Name Value pairs of current selections by the user.
	 */
	selectedAttributes: Dictionary
) => {
	const options: Record<
		string,
		Array< { label: string; value: string } | null >
	> = {};
	const attributeNames = Object.keys( attributes );
	const hasSelectedAttributes =
		Object.values( selectedAttributes ).filter( Boolean ).length > 0;

	attributeNames.forEach( ( attributeName ) => {
		const currentAttribute = attributes[ attributeName ];
		const selectedAttributesExcludingCurrentAttribute = {
			...selectedAttributes,
			[ attributeName ]: null,
		};
		// This finds matching variations for selected attributes apart from this one. This will be
		// used to get valid attribute terms of the current attribute narrowed down by those matching
		// variation IDs. For example, if I had Large Blue Shirts and Medium Red Shirts, I want to only
		// show Red shirts if Medium is selected.
		const matchingVariationIds = hasSelectedAttributes
			? getVariationsMatchingSelectedAttributes(
					attributes,
					variationAttributes,
					selectedAttributesExcludingCurrentAttribute
			  )
			: null;
		// Uses the above matching variation IDs to get the attributes from just those variations.
		const validAttributeTerms =
			matchingVariationIds !== null
				? matchingVariationIds.map(
						( varId ) =>
							variationAttributes[ 'id:' + varId ].attributes[
								attributeName
							]
				  )
				: null;
		// Intersects attributes with valid attributes.
		options[ attributeName ] = getValidSelectControlOptions(
			currentAttribute.terms,
			validAttributeTerms
		);
	} );

	return options;
};

/**
 * Return the default values of the given attributes in a format ready to be set in state.
 *
 * @return Default attributes.
 */
export const getDefaultAttributes = (
	/**
	 * List of attribute names and terms.
	 *
	 * As returned from {@link getAttributes()}.
	 */
	attributes: Record< string, ProductResponseAttributeItem >
) => {
	if ( ! isObject( attributes ) ) {
		return {};
	}

	const attributeNames = Object.keys( attributes );

	if ( attributeNames.length === 0 ) {
		return {};
	}

	const attributesEntries = Object.values( attributes );

	return attributesEntries.reduce( ( acc, curr ) => {
		const defaultValues = curr.terms.filter( ( term ) => term.default );

		if ( defaultValues.length > 0 ) {
			acc[ curr.name ] = defaultValues[ 0 ]?.slug;
		}

		return acc;
	}, {} as Dictionary );
};
