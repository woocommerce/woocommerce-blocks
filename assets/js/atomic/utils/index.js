export * from './get-block-map.js';
export * from './create-blocks-from-template.js';
export * from './render-parent-block.js';
export * from './render-inner-blocks.js';
