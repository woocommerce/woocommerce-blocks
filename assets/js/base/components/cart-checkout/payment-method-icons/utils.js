/**
 * For an array of icons, normalize into objects and remove duplicates.
 *
 * @param {Array} icons Array of icon objects or string based ids.
 */
export const normalizeIconConfig = ( icons ) => {
	const normalizedIcons = {};

	icons.forEach( ( raw ) => {
		let icon = {};

		if ( typeof raw === 'string' ) {
			icon = {
				id: raw,
				alt: raw,
				src: null,
			};
		}

		if ( typeof raw === 'object' ) {
			icon = {
				id: raw.id || '',
				alt: raw.alt || '',
				src: raw.src || null,
			};
		}

		if ( icon.id && ! normalizedIcons[ icon.id ] ) {
			normalizedIcons[ icon.id ] = icon;
		}
	} );

	return Object.values( normalizedIcons );
};
