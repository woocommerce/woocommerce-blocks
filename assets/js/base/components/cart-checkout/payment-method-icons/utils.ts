/**
 * External dependencies
 */
import type {
	PaymentMethodIcon,
	PaymentMethodIcons,
} from '@woocommerce/type-defs/payments';
import { isString } from '@woocommerce/types';

/**
 * For an array of icons, normalize into objects and remove duplicates.
 */
export const normalizeIconConfig = (
	icons: PaymentMethodIcons
): PaymentMethodIcon[] => {
	const normalizedIcons: Record< string, PaymentMethodIcon > = {};

	icons.forEach( ( raw ) => {
		let icon: Partial< PaymentMethodIcon > = {};

		if ( typeof raw === 'string' ) {
			icon = {
				id: raw,
				alt: raw,
				src: null,
			};
		}

		if ( typeof raw === 'object' ) {
			icon = {
				id: raw.id || '',
				alt: raw.alt || '',
				src: raw.src || null,
			};
		}

		if ( icon.id && isString( icon.id ) && ! normalizedIcons[ icon.id ] ) {
			normalizedIcons[ icon.id ] = <PaymentMethodIcon>icon;
		}
	} );

	return Object.values( normalizedIcons );
};
