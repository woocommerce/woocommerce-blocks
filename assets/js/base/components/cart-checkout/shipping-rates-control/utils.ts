/**
 * External dependencies
 */
import { _n, sprintf } from '@wordpress/i18n';
import { speak } from '@wordpress/a11y';

export const speakFoundShippingOptions = (
	packageCount: number,
	rateCount: number
) => {
	if ( packageCount === 1 ) {
		speak(
			sprintf(
				/* translators: %d number of shipping options found. */
				_n(
					'%d shipping option was found.',
					'%d shipping options were found.',
					rateCount,
					'woo-gutenberg-products-block'
				),
				rateCount
			)
		);
	} else {
		speak(
			sprintf(
				/* translators: %d number of shipping packages packages. */
				_n(
					'Shipping option searched for %d package.',
					'Shipping options searched for %d packages.',
					packageCount,
					'woo-gutenberg-products-block'
				),
				packageCount
			) +
				' ' +
				sprintf(
					/* translators: %d number of shipping options available. */
					_n(
						'%d shipping option was found',
						'%d shipping options were found',
						rateCount,
						'woo-gutenberg-products-block'
					),
					rateCount
				)
		);
	}
};
