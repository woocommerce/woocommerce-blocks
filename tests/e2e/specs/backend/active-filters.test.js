/**
 * External dependencies
 */
import {
	getAllBlocks,
	switchUserToAdmin,
	getEditedPostContent,
} from '@wordpress/e2e-test-utils';
import { clearAndFillInput } from '@woocommerce/e2e-utils';
import { insertBlockDontWaitForInsertClose } from '../../utils.js';
import { visitBlockPage } from '@woocommerce/blocks-test-utils';

const block = {
	name: 'Active Product Filters',
	slug: 'woocommerce/active-filters',
	class: '.wc-block-active-filters',
};

describe( `${ block.name } Block`, () => {
	beforeAll( async () => {
		await switchUserToAdmin();
		await visitBlockPage( `${ block.name } Block` );
	} );

	it( 'can be inserted once', async () => {
		await insertBlockDontWaitForInsertClose( block.name );
		expect( await getAllBlocks() ).toHaveLength( 1 );
	} );

	it( 'renders without crashing', async () => {
		await expect( page ).toRenderBlock( block );
	} );

	it( 'can change title label in editor', async () => {
		await expect( page ).toFill(
			'.wc-block-active-filters__title > textarea',
			'Current filters'
		);

		await clearAndFillInput(
			'.wc-block-active-filters__title > textarea',
			'Active filters'
		);

		expect( await getEditedPostContent() ).toMatchSnapshot();
	} );
} );
