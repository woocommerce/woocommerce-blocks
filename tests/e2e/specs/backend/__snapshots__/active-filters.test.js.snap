// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active Product Filters Block can change title label in editor 1`] = `
"<!-- wp:woocommerce/active-filters -->
<div class=\\"wp-block-woocommerce-active-filters is-loading\\" data-display-style=\\"list\\" data-heading=\\"Active filters\\" data-heading-level=\\"3\\"><span aria-hidden=\\"true\\" class=\\"wc-block-active-product-filters__placeholder\\"></span></div>
<!-- /wp:woocommerce/active-filters -->"
`;
